File: 2021/09/05/next_greater_element.py
"""
https://leetcode.com/problems/next-greater-element-i/
Author: Andrew Jarombek
Date: 9/5/2021
"""

# Back home

from typing import List


def next_greater_element(nums1: List[int], nums2: List[int]) -> List[int]:
    locations = {key: index for index, key in enumerate(nums2)}
    result = []

    for num in nums1:
        found = False

        for i in range(locations[num], len(nums2)):
            if nums2[i] > num:
                result.append(nums2[i])
                found = True
                break

        if not found:
            result.append(-1)

    return result


if __name__ == '__main__':
    assert next_greater_element([4, 1, 2], [1, 3, 4, 2]) == [-1, 3, -1]

File: 2022/09/06/nextGreaterElement.js
/**
 * https://leetcode.com/problems/next-greater-element-i/
 * @author Andrew Jarombek
 * @since 9/6/2022
 */

const nextGreaterElement = function(nums1, nums2) {
    const map = new Map();

    for (let i = 0; i < nums2.length; i++) {
        const val = nums2[i];

        let found = false;
        let index = i + 1;

        while (!found && index < nums2.length) {
            if (nums2[index] > val) {
                found = true;
            } else {
                index++;
            }
        }

        map.set(val, found ? nums2[index] : -1);
    }

    return nums1.map((val) => map.get(val));
}

let result = nextGreaterElement([4,1,2], [1,3,4,2]);
let expected = [-1,3,-1]
console.assert(expected.every((value, index) => value === result[index]));

result = nextGreaterElement([2,4], [1,2,3,4]);
expected = [3,-1];
console.assert(expected.every((value, index) => value === result[index]));
